TypeScript Adoption Plan: get it started :tada: (#5)
chriskrycho/ember-rfcs: text/0000-ts-adoption-plan.md

Where that RFC was concerned with *whether* we should pursue those goals, this RFC is concerned with *how* we accomplish them. The specific approach described here aims to roll out TypeScript support in a way that:

- maintains Ember’s strong stability guarantees
- allows us to make steady incremental progress
- communicates progress clearly to the community
- minimizes the migration costs for existing Ember TypeScript users

## Detailed design

To fully support TypeScript across the Ember ecosystem, we need:

- a Semantic Versioning policy which can absorb breaking changes in TypeScript itself *without* imposing breaking changes on Ember developers
- an edition support policy which tells us what our types *must* support (as distinct from what they *may* support)
- a plan for migrating existing Ember TypeScript users from DefinitelyTyped to using Ember’s own core types
- a plan for CLI integration (e.g. `ember new --typescript`)

> - test infrastructure for Ember’s type, both to prevent regressions and to catch breaking changes from TypeScript early so they can be mitigated

Reviewer: ```suggestion
- test infrastructure for Ember’s types, both to prevent regressions and to catch breaking changes from TypeScript early so they can be mitigated
```

- release infrastructure to allow us to handle pre-release testing, feature flags, and alpha and beta releases
- updates to our documentation to include TypeScript as a first-class citizen of the ecosystem

### Semantic Versioning

Ember packages which publish types will adhere to the [Semantic Versioning for TypeScript Types][rfc-0730] proposal. See that RFC for details on strategies for mitigating breaking changes from TypeScript, and see [**How We Teach This: Documenting SemVer**](#documenting-semver) below for discussion of how we will document this at the framework level.

The 10,000-foot summary of our treatment of SemVer for TypeScript is:

> Code which type checks at any given point in a major release of Ember packages will continue to type check throughout that Ember major release for consumers who use supported TypeScript versions.

> This is **the “no new red squiggles rule”:** if you use supported TypeScript versions, you will not see new “red squiggles” for type errors in your editor in an Ember major version.

Author: ```suggestion
This is **the “no new red squiggles rule”:** if you use supported TypeScript versions, you will not see new “red squiggles” for type errors in your editor for Ember minor version upgrades.
```

(I don't love this but haven’t come up with something I *do* love. Input welcome.)

The only exceptions to the “no new red squiggles” rule are:

- *Bug fixes to the types*: where we previously allowed code we should not have and which could produce runtime errors, the introduction of those red squiggles is making your code safer.

> - *Rare cases involving inferred types*: very rarely, we may make a change to the types which is safe for your code at runtime, and which allows TypeScript to detect new unused code paths or similar. These will introduce “red squiggles” in your editor, but in a way which will simply allow you to clean up code you didn’t need.

Reviewer: To head off any FUD here, would it be worthwhile to link to an example of this so folks can see an example of what "code you didn't need" looks like? (I think the semver RFC had one or more of those?)

Author: Good callout. I'll add some. And yeah: Ryan Cavanaugh found multiple examples of this. :joy:

````suggestion
- *Rare cases involving inferred types*: very rarely, we may make a change to the types which is safe for your code at runtime, and which allows TypeScript to detect new unused code paths or similar. These will introduce “red squiggles” in your editor, but in a way which will simply allow you to clean up code you didn’t need.

    <details><summary>example of this kind of change</summary>

    Assume a library provides a function, `idFor`, which gives back the `id` for a given item. Maybe it finds an `id` field on the object and hands it back if it exists, or generates a UUID otherwise, or similar—the details are unimportant, as we only care about the type signature.
    
    In v1.0 of the library, it specifies that it always returns a `string` or ` a `number`:

    ```ts
    declare function idFor(obj: unknown): string | number;
    ```

    In our own code, we need IDs to always be `string`s, so we do something like this:

    ```ts
    const id = idFor({ cool: 'story' });
    const usableId = typeof id === 'number' ? id.toString() : id;
    ```

    Then let’s say that the authors of `idFor` recognize that returning `string | number` is kind of annoying, and they update their API to always return `string`:

    ```ts
    function idFor(obj: unknown): string;
    ```    
    
    This is perfectly safe from a runtime perspective. Any code which worked before will work now. However, TypeScript is smart enough to see that the `typeof` check is dead code:

    ```ts
    const id = idFor({ cool: 'story' });
    const usableId = typeof id === 'number' ? id.toString() : id; // ❌
    ```

    TypeScript will now report `"Property 'toString' does not exist on type 'never'."` on the `id.toString()` call, because it determines that `id` can never be a `number`, which makes that entire branch dead code. ([playground][id-play]) The "cleanup" work here is that you can simply *get rid of* that code, though!

    For further examples and a discussion of why we cannot work around this (e.g. with lint rules), see [**Appendix C**](https://github.com/chriskrycho/ember-rfcs/blob/semver-for-ts/text/0730-semver-for-ts.md#appendix-c-on-variance-in-typescript) in the Semantic Versioning for TypeScript Types RFC.

    </details>

[id-play]: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEBbA9sArhBA1AjPA3gFDwnwgAeADsjAC7wBmaAdmLQJbLPzvABiNABTIARgCsAXPBYBrZsgDuzAJRSAzrRjtmAc3gAfeMzSIRIGAG5S8QgF9ChUJFgIU6TPCwAmAsVIVqOkYWNk5uXgEYYXEpWXklVXgNLV0LOwcmVg4ueAA3HEFlX2swLg0eYHgAXk8cADoIoXx4UuQIKQByDRoATw74W2U0krL6NDUoEUwASUqa2h7KEGQGCuqqmo7jU3N+gH4KutpkAGVNbR1C+CleNPtCTNCc3K8rohHmct5qzy8G-iaLWQbU63RgfQGQz8JFKnzGEymIFmPwWSxWaw2m22Zhg+0OxzOKUuRRuwDuhCAA

````

For details on both of these, see [Semantic Versioning for TypeScript Types RFC][rfc-0730], which explains these in detail and provides justification for especially the second case.

[rfc-0730]: https://github.com/emberjs/rfcs/pull/730

#### Strictness

> In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultExports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)

Reviewer: ```suggestion
In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultImports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)
```

> In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultImports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)

Author: ```suggestion
In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], both Ember projects and the blueprints generated by Ember using the new `--typescript` flag will set the following compiler options:

- `strict: true`
- `noUncheckedIndexedAccess: true`
- `esModuleInterop: false`
- `allowSyntheticDefaultImports: false`

(See discussion below under [**CLI Integration**](#cli-integration) for further details on the new `--typescript` flag`.)
```

Beyond this, we *may* define a set of *additional* “lint”-type strictness flags over time taking advantage of TypeScript and/or ESLint, an `ember-ultra-strict` mode, and provide tooling to expose that.

> A consequence of adopting `strict: true` as a behavior is that we will be progressively catching *more* type safety issues over time. Since these changes are not gated on TypeScript minor versions, they make cause “red squiggles” during any TypeScript upgrade. However, we still believe this is an appropriate default, for the following reasons:

Author: ```suggestion
A consequence of adopting `strict: true` as a behavior is that we will be progressively catching *more* type safety issues over time. Since these changes are not gated on TypeScript major versions (that is, they may appear in *any* TypeScript release), they make cause “red squiggles” during any TypeScript upgrade. However, we still believe this is an appropriate default, for the following reasons:
```

<details><summary>On including <code>noUncheckedIndexedAccess</code></summary>

We include the `noUncheckedIndexedAccess` setting in our recommended settings because it catches one of the most significant holes in the TypeScript type system prior to its introduction: not accounting for the fact that "index"-style access to objects and arrays *always* allows arbitrary access to any index (numeric or string), whether or not it is defined. For example, consider this type (not recommended for *other* reasons, but legal):

```ts
type Anything {
  [key: string]: string;
}

> function describe(anything: anything) {

Reviewer: ```suggestion
function describe(anything: Anything) {
```

  let email = person['email'];
}
```

Here, `email` would be typed as `string`, and so TypeScript would not complain if you did something like `email.length`. But you can *call* this with an object which does *not* have an `email` field on it:

```ts
describe({ noEmailsHere: 'bwahahaha' });
```

This will cause a runtime error instead of a type error. However, TypeScript is (as of recent releases) smart enough to understand that if you *check* for `"email"` it can be *known* to be present:

```ts
function describe(anything: Anything) {
  if (anything['email']) {
    let email = anything['email']
    email.length
  }
}
```

Accordingly, we include this as part of Ember’s own strictness settings and defaults generated for users.[^should-be-strict]

</details>

[^should-be-strict]: The Typed Ember team thinks TypeScript *should* have included this under `strict: true`!

#### Primary packages

> The primary packages in the ecosystem, `ember-source`, `ember-data`, and `ember-cli`, will adopt the [rolling window](https://github.com/chriskrycho/ember-rfcs/blob/semver-for-ts/text/0730-semver-for-ts.md#rolling-support-windows) policy:

Reviewer: In a future where these packages are broken out into their constituents (which is already partly true with the `@ember-data/*` packages that ship today), is there an easy way as a user to know what's considered a primary package and what isn't?

Author: Hmmmmmmmmm. This is a very interesting question. I don’t have any *great* thoughts on it, to be honest. I think it’s probably a thing we should tackle if/as that happens? But I could also be persuaded we need to address it here. 🤔 Going to add it to **Unresolved questions**.

Author: On reflection as I started writing it up, I instead integrated it into **Alternatives: Documentation**. Thank you for calling it out!

In future editions, it is conceivable that we might lean on this escape hatch to drop support for legacy CLI features (after Embroider is standard) or legacy routing features (if Ember adopts a modernized routing system) etc. However, in general we do *not* expect to need it, because our designs have generally been accounting for TypeScript even before making it an officially supported language for the framework.

### RFC process updates

RFCs which introduce new APIs must specify the public types for those APIs. RFCs which deprecate existing APIs must consider the migration path for TypeScript users as well as JavaScript users.

Note that this is not a major new design constraint, because we have largely already been doing this! This simply formalizes the process Ember has been using for the past few years already: we have regularly used TypeScript signatures to represent new APIs, and TypeScript concerns have been included in the discussion of API proposals.

This does introduce two new small considerations:

> - New APIs will now need to specify import paths for type-only imports. For example, today there is no public import for `Transition` or `RouteInfo`, because they are not intended for end users to construct. However, types for them *should* be publicly exported because end users need to be able to name those types them in their own code. If we were to write [RFC #0095: Router Service](https://emberjs.github.io/rfcs/0095-router-service.html) today, we would specify imports for both.

Reviewer: ❤️

To this end, we will extend our existing test infrastructure in the following ways:

1. We will convert our test suite to use TypeScript, so that all existing tests can be used for catching type errors.
2. We will add dedicated “type tests” using tooling like [expect-type](https://github.com/mmkal/ts/tree/master/packages/expect-type) to guarantee we are upholding *exactly* the contract we intend for our published types.
3. We will add CI jobs which type check the code against all supported TypeScript versions.
4. We will incorporate checking against the `typescript@next` nightly builds so that we can identify breaking changes early.

Point (4) will make sure that we have early warning about any inbound breakage from TypeScript itself, generally giving us on the order of 2–3 Ember minor release cycles to prepare for those changes and prevent them from breaking Ember users.

### Publishing types and release channels

> When Ember official packages begin publishing types, they will need to account for TypeScript's default of using Node module resolution to lookup. In the future, we will be able to publish types in a much more 

Reviewer: Missing the end of this sentence?

Author: Missing most of a paragraph, in fact. :joy: Updated!

[^transition-on-DT]: Historically, users on DefinitelyTyped have had to import `Transition` and `RouteInfo` from private locations *or* use type hacks to get them from the return types of other functions. This is because Typed Ember has consistently maintained a rule that we only publish types corresponding to the documented public API of Ember itself, and there was no possible public API for these types as *type-only* exports, since Ember was not publishing types.

### Migration docs

As discussed briefly above, existing Ember TypeScript users have used the `@types` packages from [DefinitelyTyped][DT]. We must document a migration path for users, to be published as part of the blog post announcing officially-published types for any given package. This will include both a general migration path, common to all of these, and then specific details dealing with migration of specific Ember Classic patterns.

#### General migration guide

The outline of that migration path, regardless of the specific package, is:

1. Upgrade to a version of the package (e.g. `ember-source`) which officially publishes types.

> 2. Remove the corresponding `@types` package(s) from.

Reviewer: ```suggestion
2. Remove the corresponding `@types` package(s) from your dependencies.
```
